src/main/java/com/zendesk/maxwell/MaxwellUpdateRowsEvent.java

```diff
@@ -48,4 +52,64 @@ public String sqlOperationString() {
 	public String getType() {
 		return "update";
 	}
+
+	private LinkedList<Pair<Row>> filteredRowsBeforeAndAfter;
+	private boolean performedBeforeAndAfterFilter;
+
+	private List<Pair<Row>> filteredRowsBeforeAndAfter() {
+		if ( this.filter == null)
+			return event.getRows();
+
+		if ( performedBeforeAndAfterFilter )
+			return filteredRowsBeforeAndAfter;
+
+		filteredRowsBeforeAndAfter = new LinkedList<>();
+		for ( Pair<Row> p : event.getRows()) {
+			if ( this.filter.matchesRow(this, p.getAfter()) )
+				filteredRowsBeforeAndAfter.add(p);
+		}
+		performedBeforeAndAfterFilter = true;
+		return filteredRowsBeforeAndAfter;
+	}
+
+	@Override
+	public List<RowMap> jsonMaps() {
+		ArrayList<RowMap> list = new ArrayList<>();
+		Object avalue;
+		Object bvalue;
+		for (Pair<Row> p : filteredRowsBeforeAndAfter() ) {
+			Row after = p.getAfter();
+			Row before = p.getBefore();
+
+			RowMap rowMap = buildRowMap();
+
+			Iterator<Column> aftIter = after.getColumns().iterator();
+			Iterator<Column> befIter = before.getColumns().iterator();
+			Iterator<ColumnDef> defIter = table.getColumnList().iterator();
+			while ( aftIter.hasNext() && defIter.hasNext() && befIter.hasNext() ) {
+				Column afterColumn = aftIter.next();
+				ColumnDef columnDef = defIter.next();
+				Column beforeColumn = befIter.next();
+
+				avalue = valueForJson(afterColumn);
+
+				bvalue = valueForJson(beforeColumn);
+
+				if (avalue != null) {
+					avalue = columnDef.asJSON(avalue);
+					if ( !avalue.equals(columnDef.asJSON(bvalue)) ) {
```

Comment:
nope, still too complex.  Really, just try the code as 3 if statements:

```
  if ( avalue != null && bvalue != null ) { 
    // run comparison
  } else if ( bvalue != null ) { 
     // avalue is null, bvalue is not -- cast bvalue to json and put in oldData
  }  else if ( avalue != null ) { 
     // bvalue is null, avalue is not -- put null in oldData
 } // otherwise, both are null, do nothing
```